src/shared/components/data-table/components/table-controls/components/table-actions/components/download-data/DownloadData.tsx

```diff
@@ -0,0 +1,134 @@
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React, { ReactNode, useEffect, useRef } from 'react';
+import ReactDOM from 'react-dom/client';
+import memoize from 'lodash/memoize';
+
+import useDataTable from 'src/shared/components/data-table/hooks/useDataTable';
+import { downloadTextAsFile } from 'src/shared/helpers/downloadAsFile';
+
+import LoadingButton from 'src/shared/components/loading-button';
+
+import { TableAction } from 'src/shared/components/data-table/dataTableTypes';
+
+import styles from './DownloadData.scss';
+
+const getReactRenderer = memoize(() => {
+  const element = document.createElement('div');
+  const root = ReactDOM.createRoot(element);
+
+  return {
+    element,
+    renderer: root
+  };
+});
+
+const getReactNodeText = (node: ReactNode): string => {
+  const { element, renderer } = getReactRenderer();
+  renderer.render(node);
+  return element.innerText;
+};
+
+const DownloadData = () => {
+  const {
+    dispatch,
+    rows,
+    downloadFileName,
+    columns,
+    selectedAction,
+    hiddenRowIds,
+    downloadHandler
+  } = useDataTable();
+
+  const allowComponentResetRef = useRef(true);
+
+  useEffect(() => {
+    return () => {
+      allowComponentResetRef.current = false;
+    };
+  }, []);
+
+  const restoreDefaults = () => {
+    setTimeout(() => {
+      if (allowComponentResetRef.current) {
+        dispatch({
+          type: 'set_selected_action',
+          payload: TableAction.DEFAULT
+        });
+      }
+    }, 1000);
+  };
+
+  const handleDownload = async () => {
+    if (downloadHandler) {
+      await downloadHandler();
+      setTimeout(restoreDefaults, 1000);
+      return;
+    }
+    const dataForExport: string[][] = [];
+    dataForExport[0] = [
+      ...columns
+        .filter((column) => column.isExportable !== false)
+        .map((column) => column.title ?? '')
+    ];
+
+    const rowsToDownload =
+      selectedAction === TableAction.DOWNLOAD_ALL_DATA
+        ? rows
+        : rows.filter((row) => !hiddenRowIds[row.rowId]);
+
+    rowsToDownload.forEach((row, rowIndex) => {
+      dataForExport[rowIndex + 1] = [];
+      row.cells.forEach((cell, cellIndex) => {
+        const { renderer, isExportable } = columns[cellIndex];
+
+        if (isExportable !== false) {
+          const cellExportData = renderer
+            ? renderer({
+                rowData: row.cells,
+                rowId: String(row.rowId),
+                cellData: cell
+              })
+            : cell;
+
+          dataForExport[rowIndex + 1].push(getReactNodeText(cellExportData));
```

Comment:
Don't you want to check here if `cellExportData` is a primitive (string / number), in which case you could push it into `dataForExport` directly?

Reply:
That's done. Previously it was getting checked inside `getReactNodeText` but not since we updated it.